modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java

```diff
@@ -187,20 +232,33 @@ public GridEncryptionManager(GridKernalContext ctx) {
         }, EVT_NODE_LEFT, EVT_NODE_FAILED);
 
         ctx.io().addMessageListener(TOPIC_GEN_ENC_KEY, ioLsnr = (nodeId, msg, plc) -> {
-            synchronized (genEcnKeyMux) {
+            synchronized (opsMux) {
                 if (msg instanceof GenerateEncryptionKeyRequest) {
                     GenerateEncryptionKeyRequest req = (GenerateEncryptionKeyRequest)msg;
 
                     assert req.keyCount() != 0;
 
                     List<byte[]> encKeys = new ArrayList<>(req.keyCount());
 
-                    for (int i = 0; i < req.keyCount(); i++)
-                        encKeys.add(getSpi().encryptKey(getSpi().create()));
+                    byte[] masterKeyDigest;
+
+                    masterKeyChangeLock.readLock().lock();
```

Comment:
Can we made a utility method for executing login under read lock?
```
private void withReadLock(Runnable method) {
        masterKeyChangeLock.readLock().lock();

        try {
            logic.run();
        }
        finally {
            masterKeyChangeLock.readLock().unlock();
        }
}
```

Reply:
Done. I have used `Callable` instead of `Runnable`. Please, take a look.